apps/prairielearn/src/lib/externalGraderCommon.js

```diff
@@ -21,169 +20,91 @@ export function getJobDirectory(jobId) {
-              if (ERR(err, callback)) return;
-              callback(null);
-            });
-          } else {
-            logger.warn(
-              `No tests directory found for ${question.qid}; maybe you meant to specify some?`,
-            );
-            callback(null);
-          }
-        });
-      },
-      (callback) => {
-        if (submission.submitted_answer._files) {
-          async.each(
-            submission.submitted_answer._files,
-            (file, callback) => {
-              if (!file.name) {
-                return callback(new Error("File was missing 'name' property."));
-              }
-              if (file.contents == null) {
-                return callback(new Error("File was missing 'contents' property."));
-              }
+  try {
+    // Attempt to remove existing directory first
+    await fsPromises.rm(dir, { force: true, recursive: true });
+    await fsPromises.mkdir(dir, { recursive: true });
+    await fsPromises.mkdir(path.join(dir, 'serverFilesCourse'));
+    await fsPromises.mkdir(path.join(dir, 'tests'));
+    await fsPromises.mkdir(path.join(dir, 'student'));
+    await fsPromises.mkdir(path.join(dir, 'data'));
+
+    // Copy all specified files/directories into serverFilesCourse/
+    for (const file of question.external_grading_files ?? []) {
+      const src = path.join(coursePath, 'serverFilesCourse', file);
+      const dest = path.join(dir, 'serverFilesCourse', file);
+      await fsExtra.copy(src, dest);
+    }
 
-              // Files are expected to be base-64 encoded
-              let decodedContents = Buffer.from(file.contents, 'base64');
-              // Check that the file name does not try to navigate up in
-              // the directory hierarchy
-              const fullPath = path.join(dir, 'student', file.name);
-              if (!contains(path.join(dir, 'student'), fullPath)) {
-                callback(new Error('Invalid filename'));
-                return;
-              }
-              // The version of ensureDir provided by fs-extra
-              // automatically creates intermediate paths recursively.
-              const targetDir = path.dirname(fullPath);
-              fs.ensureDir(targetDir, (err) => {
-                if (ERR(err, callback)) return;
-                fs.writeFile(fullPath, decodedContents, (err) => {
-                  if (ERR(err, callback)) return;
-                  callback(null);
-                });
-              });
-            },
-            function (err) {
-              if (ERR(err, callback)) return;
-              callback(null);
-            },
+    // This is temporary while /grade/shared is deprecated but still supported
+    // TODO remove this when we remove support for /grade/shared
+    const src = path.join(dir, 'serverFilesCourse');
+    const dest = path.join(dir, 'shared');
+    await fsExtra.copy(src, dest);
+
+    if (question.directory != null) {
```

Comment:
This test was added because the type of `question.directory` is nullable, though I would expect this not to be the case in practice.